0012-unicode-strings.md

```diff
@@ -0,0 +1,445 @@
+away from encodings, especially to support correct code that must compile to multiple target languages.
+Providing more of the concepts defined by the Unicode standard is left out of scope for this proposal,
+under the assumption that it will enable such implementations via Dafny source code in sharable libraries instead.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Verification
+
+The change to the definition of `char` in the `DafnyPrelude.bpl` file is minimal.
+The only change is to replace expressions such as `0 <= n && n < 65536` with a separate predicate defined as follows:
+
+```boogie
+function char#IsUnicodeScalarValue(n: int): bool {
+  (0                  <= n && n < 55296   /* 0xD800 */) || 
+  (57344 /* 0xE000 */ <= n && n < 1114112 /* 0x11_0000 */)
+}
+```
+
+As described above in the user-facing description of this feature, this will cause verification failures by design.
+Unlike many other programming languages that have added Unicode support over a major version bump,
+Dafny is unusually well-positioned to catch most regressions statically.
+
+## Compilation/Runtime
+
+This change also has a reasonably small effect on the runtime behavior of any self-contained Dafny program,
+meaning a program that does not pass data over an interface with external code.
+The only difference is that the `char` type will need to be compiled to a native datatype
+that supports 32 bits of data (or technically 21 according to the actual Unicode scalar value range)
+rather than 16 bits as previous versions did.
+Many target languages will have dedicated types designed to hold any Unicode scalar value,
+such as the `char` primitive type in Rust or the `rune` alias for `int32` in Go.
+
+The main complexity in implementing this change is minimizing the migration pain for
+code bases that use `{:extern}` definitions to pass string values in and out of the Dafny runtime.
+This generally requires converting between the internal representation of Dafny strings
+and the native implementation in the target language,
+e.g. between `Dafny.ISequence<char>` and `System.String` in C#.
+Even there, the various Dafny runtime libraries provide helper methods that can be adjusted 
+to work with the new semantics of Unicode characters.
+In the C# runtime, for example, the `Sequence.FromString(string)` method converts a native string
+to an equivalent `Dafny.ISequence<char>` copy.
+A parallel method named something similar to `Sequence.UnicodeFromString(string)` could be added
+that would return a `Dafny.ISequence<uint32>` copy instead.
+Migrating an existing codebase should reduce to a simple find-and-replace operation.
+
+Note that this aspect of the foreign function interface in Dafny is best handled on a per-backend basis.
+It isn't generally tractable to define a single `NativeString` type with consistent semantics,
```

Comment:
Could you specify in more details what the limitations of a `NativeString` type would be?

I imagine it would be a subset of a collection type with linear time access, like 

```type NativeString = x: LinkedList<char> | ValidInAllRuntimes(x)```

The LinkedList would were users against doing random access on individual characters, since that requires decoding depending on the runtime.